admission_control/admission-control-service/src/upstream_proxy.rs

```diff
@@ -420,17 +420,27 @@ where
         add_transaction_request.latest_sequence_number = sequence_number;
     }
 
-    add_txn_to_mempool(&upstream_proxy_data, add_transaction_request)
+    add_txn_to_mempool(&upstream_proxy_data, add_transaction_request).await
 }
 
-fn can_send_txn_to_mempool<M, V>(upstream_proxy_data: &UpstreamProxyData<M, V>) -> Result<bool>
+async fn can_send_txn_to_mempool<M, V>(
+    upstream_proxy_data: &UpstreamProxyData<M, V>,
+) -> Result<bool>
 where
     M: MempoolClientTrait,
 {
     if upstream_proxy_data.need_to_check_mempool_before_validation {
         let req = HealthCheckRequest::default();
         let is_mempool_healthy = match &upstream_proxy_data.mempool_client {
-            Some(client) => client.health_check(&req)?.is_healthy,
+            Some(client) => {
+                client
+                    .as_ref()
+                    .clone()
+                    .health_check(req)
+                    .await?
+                    .into_inner()
+                    .is_healthy
```

Comment:
This is quite a bit worse than it was. Are there other options that don't make using client so tedious?

Reply:
Yeah it is a lot worse due to lots of our existing code relying on the API that was provided by grpcio. As an example, all current RPC method calls take immutable references whereas tonic requires &mut refs. This causes some difficulty because we use Arc<dyn Trait> objects everywhere which is inherently incompatible with taking a &mut ref. This was a (poor) workaround to that issue, thought maybe we need to go further and do some more aggressive refactoring first?